File: src/IcsiBoost.java
/* Copyright (C) (2009) (Benoit Favre) <favre@icsi.berkeley.edu>

This program is free software; you can redistribute it and/or 
modify it under the terms of the GNU Lesser General Public License 
as published by the Free Software Foundation; either 
version 2 of the License, or (at your option) any later 
version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA. */

import java.util.regex.*;
import java.util.*;
import java.io.*;


class IcsiBoost {
    public class Feature {
    }

    public class ContinuousFeature extends Feature {
        double value;
        public ContinuousFeature(String content) {
            value = Double.parseDouble(content);
        }
        public double getValue() {
            return value;
        }
	public String toString() {
	    return String.format("%f", value);
	}

    }

    public class TextFeature extends Feature {
        protected List<String> values;
        public TextFeature() { 
            values = new ArrayList<>();
        }
        public TextFeature(String content) {
            values = new ArrayList<>(1);
            values.add(content);
        }
        public List<String> getValues() {
            return values;
        }
    }

    public class NGramFeature extends TextFeature {
        public NGramFeature(String content, int length) {
            values = new ArrayList<>();
            String[] tokens = content.split(" ");
            for(int i = 0; i < tokens.length; i++) {
                StringBuilder ngram = new StringBuilder();
		// JHE
                //System.out.println("zzzz"+ tokens[i] +" " + i + " " + length);
                for(int j = 0; j < length && i + j < tokens.length; j++) {
                    if (ngram.length() != 0) ngram.append("#");
                    ngram.append(tokens[i + j]);
		values.add(ngram.toString());
                }
            }
        }
	public String toString() {
	    String r = "<" + values + ">";
            return r;
	}
    }

    public class FGramFeature extends TextFeature {
        public FGramFeature(String content, int length) {
            values = new ArrayList<>();
            String[] tokens = content.split(" ");
            for(int i = 0; i < tokens.length - length; i++) {
                StringBuilder ngram = new StringBuilder(tokens[i]);
                for(int j = 0; j < length; j++) {
                    ngram.append("_");
                    ngram.append(tokens[i + j]);
                }
                values.add(ngram.toString());
            }
        }
	public String toString() {
	    String r = "<" + values + ">";
            return r;
	}

    }

    public class SGramFeature extends TextFeature {
        public SGramFeature(String content, int length) {
            values = new ArrayList<>();
            String[] tokens = content.split(" ");
            for(int i = 0; i < tokens.length - length; i++) {
                values.add(tokens[i] + "_" + tokens[i + length]);
            }
        }
	public String toString() {
	    String r = "<" + values + ">";
            return r;
	}

    }

    public class UndefinedFeature extends Feature {
    }

    public class Classifier {
        protected double alpha;
        protected int column;
        protected double c0[];
        protected double c1[];
        protected double c2[];
        public void classify(Feature feature, double[] result){ }
        public String toString() {
            StringBuilder builder = new StringBuilder(alpha + " " + column);
            builder.append("\n");
            if(c0 != null) {
                builder.append("  c0 ");
                for(int i = 0; i < c0.length; i++) {
                    builder.append(c0[i]).append(" ");
                }
                builder.append("\n");
            }
            if(c1 != null) {
                builder.append("  c1 ");
                for(int i = 0; i < c1.length; i++) {
                    builder.append(c1[i]).append(" ");
                }
                builder.append("\n");
            }
            if(c2 != null) {
                builder.append("  c2 ");
                for(int i = 0; i < c2.length; i++) {
                    builder.append(c2[i]).append(" ");
                }
                builder.append("\n");
            }
            return builder.toString();
        }
    }

    public class ThresholdClassifier extends Classifier {
        double threshold;
        public String toString() {
            String output = super.toString();
            return "threshold:" + threshold + ":" + output;
        }

        public ThresholdClassifier(double threshold, int column, double alpha, double[] c0, double[] c1, double[] c2) {
            this.threshold = threshold;
            this.column = column;
            this.alpha = alpha;
            this.c0 = c0;
            this.c1 = c1;
            this.c2 = c2;
        }

        public void classify(Feature feature, double[] result) { 
            if(UndefinedFeature.class.isInstance(feature)) {
                for(int i = 0; i < result.length; i++) {
                    result[i] += c0[i];
                }
            } else {
                double value = ((ContinuousFeature) feature).getValue();
                if(value < threshold) {
                    for(int i = 0; i < result.length; i++) {
                        result[i] += c1[i];
                    }
                } else {
                    for(int i = 0; i < result.length; i++) {
                        result[i] += c2[i];
                    }
                }
            }
        }
    }

    public class TextClassifier extends Classifier {
        String type = "text";
        String token;

        public String toString() {
            String output = super.toString();
            return "text:" + token + ":" + output;
        }

        public TextClassifier(String token, int column, double alpha, double[] c0, double[] c1) {
            this.token = token;
            this.column = column;
            this.alpha = alpha;
            this.c0 = c0;
            this.c1 = c1;
        }

        public void classify(Feature feature, double[] result) {
            if(UndefinedFeature.class.isInstance(feature)) {
                for(int i = 0; i < result.length; i++) {
                    result[i] += c0[i];
                }
            } else {
                List<String> values = ((TextFeature)feature).getValues();
                if(values.contains(token)) {
                    for(int i = 0; i < result.length; i++) {
                        result[i] += c1[i];
                    }
                } else {
                    for(int i = 0; i < result.length; i++) {
                        result[i] += c0[i];
                    }
                }
            }
        }
    }
    public class Model {
        public List<Classifier> classifiers;
        public String[] labels;
        public List<String> types;
        public Map<String, Integer> mapping;
        static final int TYPE_CONTINUOUS = 0;
        static final int TYPE_TEXT = 1;
        static final int TYPE_IGNORE = 1;
        int[] typesAsInt;
        static public final int TYPE_SGRAM = 2;
        static public final int TYPE_FGRAM = 1;
        static public final int TYPE_NGRAM = 0;
        int ngramLength = 0;
        int ngramType = 0;

        public Model(String stem, int ngramLength, int ngramType) throws IOException {
            this.ngramLength = ngramLength;
            this.ngramType = ngramType;
            loadNames(stem + ".names");
            loadShyp(stem + ".shyp");
        }
        public void loadNames(String filename) throws IOException {
            mapping = new HashMap<>();
            types = new ArrayList<>();
            boolean firstLine = true;
            BufferedReader input = new BufferedReader(new FileReader(filename));
            String line;
            while(null != (line = input.readLine())) {
                line = line.trim();
                if(line.matches("^(\\s*|\\s*\\|.*)$")) continue;
                if(firstLine) {
                    labels = line.split("(^\\s+|\\s*,\\s*|\\s*\\.?$)");
                    firstLine = false;
                } else {
                    String result[] = line.split("(^\\s+|\\s*:\\s*|\\s*\\.?$)");
                    //System.err.println("QQQQ " + line + " " + result[1] + " " + result.length);
                    if(result.length >= 2) { // etait == 2
                        mapping.put(result[0], types.size());
			//System.err.println("QQQQ " + line + " " + result[0] + " " + types.size());
                        types.add(result[1]);
                    }
                }
            }
            typesAsInt = new int[types.size()];
            for(int i = 0; i < types.size(); i++) {
                if("ignore".equals(types.get(i))) {
                    typesAsInt[i] = TYPE_IGNORE;
                } else if("continuous".equals(types.get(i))) {
                    typesAsInt[i] = TYPE_CONTINUOUS;
                } else {
                    typesAsInt[i] = TYPE_TEXT;
                }
            }
        }

        public void loadShyp(String filename) throws IOException {
            classifiers = new ArrayList<>();
            boolean seenIterations = false;
            BufferedReader input = new BufferedReader(new FileReader(filename));
            String line;
            double alpha = Double.NaN;
            String name;
            String token = null;
            double threshold;
            double c0[] = null;
            double c1[] = null;
            double c2[] = null;
            int column = -1;
            while(null != (line = input.readLine())) {
                line = line.trim();
                if(line.length() == 0) continue;
                if(!seenIterations) {
                    numIterations = Integer.parseInt(line);
                    seenIterations = true;
                } else {
                    Pattern textPattern = Pattern.compile("^\\s*(\\S+)\\s+Text:SGRAM:([^:]+):(.*?) *$");
                    Matcher textMatcher = textPattern.matcher(line);
                    Pattern thresholdPattern = Pattern.compile("^\\s*(\\S+)\\s+Text:THRESHOLD:([^:]+):");
                    Matcher thresholdMatcher = thresholdPattern.matcher(line);
                    if(textMatcher.matches()) {
                        alpha = Double.parseDouble(textMatcher.group(1));
                        name = textMatcher.group(2);
                        token = textMatcher.group(3);
                        Integer columnInteger = mapping.get(name);
                        if(columnInteger != null) {
                            column = columnInteger;
                            String type = types.get(column);
                            if(type.contains(",")) {
                                String values[] = type.split("\\s*,\\s*");
                                int tokenId = Integer.parseInt(token);
                                if(tokenId < 0 || tokenId >= values.length) {
                                    System.err.println("ERROR: value not found \"" + tokenId + "\" in names file");
                                    System.exit(1);
                                }
                                token = values[Integer.parseInt(token)];
                            }
                        } else {
                            System.err.println("ERROR: name not found \"" + name + "\" in names file");
                            System.exit(1);
                        }
                        c0 = null;
                        c1 = null;
                        c2 = null;
                    } else if(thresholdMatcher.matches()) {
                        alpha = Double.parseDouble(thresholdMatcher.group(1));
                        name = thresholdMatcher.group(2);
                        Integer columnInteger = mapping.get(name);
                        if(columnInteger != null) {
                            column = columnInteger;
                            String type = types.get(column);
                            if(!"continuous".equals(type)) {
                                System.err.println("ERROR: unsupported type \"" + type + "\" for threshold");
                                System.exit(1);
                            }
                        } else {
                            System.err.println("ERROR: name not found \"" + name + "\" in names file");
                            System.exit(1);
                        }
                        c0 = null;
                        c1 = null;
                        c2 = null;
                        token = null;
                    } else {
                        String values[] = line.split(" ");
                        if(values.length == labels.length) {
                            if(c0 == null) {
                                c0 = new double[labels.length];
                                for(int i = 0; i < values.length; i++) {
                                    c0[i] = Double.parseDouble(values[i]);
                                }
                            } else if(c1 == null) {
                                c1 = new double[labels.length];
                                for(int i = 0; i < values.length; i++) {
                                    c1[i] = Double.parseDouble(values[i]);
                                }
                                if(token != null) {
                                    TextClassifier classifier = new TextClassifier(token, column, alpha, c0, c1);
                                    classifiers.add(classifier);
                                }
                            }
                            else {
                                c2 = new double[labels.length];
                                for(int i = 0; i < values.length; i++) {
                                    c2[i] = Double.parseDouble(values[i]);
                                }
                            }
                        } else if(values.length == 1) {
                            threshold = Double.parseDouble(line);
                            ThresholdClassifier classifier = new ThresholdClassifier(threshold, column, alpha, c0, c1, c2);
                            classifiers.add(classifier);
                        } else {
                            System.err.println("ERROR: unexpected line \"" + line + "\" in shyp file");
                            System.exit(1);
                        }
                    }
                }
            }
        }
        public int numIterations;
        public Example readExample(String line) {
            Example example = new Example();
            String fields[] = line.split("\\s*,\\s*");
            if(fields.length == typesAsInt.length || fields.length == typesAsInt.length + 1) {
                for(int i = 0; i < typesAsInt.length; i++) {
                    if("?".equals(fields[i])) {
                        example.features.add(new UndefinedFeature());
                    } else {
                        if(typesAsInt[i] == TYPE_TEXT) {
                            switch (ngramType) {
                                case TYPE_NGRAM:
                                    example.features.add(new NGramFeature(fields[i], ngramLength));
                                    break;
                                case TYPE_FGRAM:
                                    example.features.add(new FGramFeature(fields[i], ngramLength));
                                    break;
                                case TYPE_SGRAM:
                                    example.features.add(new SGramFeature(fields[i], ngramLength));
                                    break;
                                default:
                                    break;
                            }
                        } else if(typesAsInt[i] == TYPE_CONTINUOUS) {
                            example.features.add(new ContinuousFeature(fields[i]));
                        }
                    }
                }
                if(fields.length == typesAsInt.length + 1) {
                    String llabels[] = fields[fields.length - 1].split("(\\s+|\\s*\\.$)");
                    for(int i = 0; i < llabels.length; i++) {
                        example.labels.add(llabels[i]);
                    }
                }
            }
	    //System.err.println("Ex.: " + example);
            return example;
        }
    }
    public class Example {
        public List<Feature> features = new ArrayList<>();
        public List<String> labels = new ArrayList<>();

	// put the results of the evaluation
        public List<Integer> haslabels = new ArrayList<>();
        public List<Double> values = new ArrayList<>();
	
	public String toString() {
	    String r = features + ":" + labels + " " + haslabels + " " + values;
	    return r;
	}
	
	public String results() {
	    StringBuilder sb = new StringBuilder();
	    boolean first = true;
	    for (Integer i : haslabels) {
		if (!first) sb.append(' ');
		else first = false;
		sb.append(i);
	    }
	    for (Double d : values) {
		sb.append(' ');
		sb.append(d);
	    }

	    return sb.toString();
	}
	
    }

    Model model;

    public Example decode(String line) {
        Example example = model.readExample(line);
        double result[] = new double[model.labels.length];
        for(int i = 0; i < model.numIterations; i++) {
            Classifier classifier = model.classifiers.get(i);
            classifier.classify(example.features.get(classifier.column), result);
        }
        for (int i = 0; i < model.labels.length; i++) {
	    //if (i != 0) System.out.print(" ");
            //if (example.labels.contains(model.labels[i])) System.out.print("1");
            //else System.out.print("0");

            if (example.labels.contains(model.labels[i])) example.haslabels.add(1);
            else example.haslabels.add(1);
	    
        }

        for(int i = 0; i < result.length; i++) {
	    //System.out.println("eeeeee " + model.numIterations + " " + result[i]);
	    double x = result[i];
	    double res = 1.0 / (1.0 + Math.exp(-2.0 * x));
	    //System.out.print(" " + res);
	    example.values.add(res);
        }
        //System.out.println();
        //System.out.println("EEE: " + example);
        return example;
    }
    public IcsiBoost(String stem) throws IOException {
        model = new Model(stem, 1, Model.TYPE_NGRAM);
    }
    public static void main(String args[]) {
        try {
            IcsiBoost decoder = new IcsiBoost(args[0]);
            BufferedReader reader = new BufferedReader(new FileReader(args[1]));
            String line;
            while(null != (line = reader.readLine())) {
                Example example = decoder.decode(line);
		System.out.println(example.results());
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/icsiboost.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) (2009) (Benoit Favre) <favre@icsi.berkeley.edu>
#
# This program is free software; you can redistribute it and/or 
# modify it under the terms of the GNU Lesser General Public License 
# as published by the Free Software Foundation; either 
# version 2 of the License, or (at your option) any later 
# version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA. */

import sys, re, math

# this script implements the prediction side of icsiboost in python
# supports continuous features and ngrams (see the main)
# sgrams and fgrams are not supported yet, but easy to implement (see generate_ngram())

class Feature:
    def __init__(self, alpha, column, value=None, threshold=None):
        self.alpha = alpha
        self.column = column
        self.value = value
        self.threshold = threshold
        self.c0 = []
        self.c1 = []
        self.c2 = []

class Classifier:
    def __init__(self, stem=None):
        self.features = []
        self.names = {}
        self.classes = []
        if stem:
            self.load_model(stem)

    def load_model(self, stem):
        self.read_names(stem + ".names")
        self.read_shyp(stem + ".shyp")

    def read_names(self, names_file):
        classes = None
        columns = []
        for line in open(names_file).xreadlines():
            line = line.strip()
            if classes == None:
                classes = [x.strip() for x in re.sub('\.$', '', line).split(',')]
                self.classes = classes
            else:
                name = re.split(r':', line)[0]
                self.names[name] = len(self.names)
        #print "NAMES", self.names
    def read_shyp(self, shyp_file):
        num_classifiers = None
        feature = None
        line_num = 0
        for line in open(shyp_file).xreadlines():
            line_num += 1
            line = line.strip()
            if line == "":
                continue
            tokens = line.split()
            found_sgram = re.search(r'^\s*(\S+)\s+Text:SGRAM:([^:]+):(.*?) *$', line)
            found_threshold = re.search(r'^\s*(\S+)\s+Text:THRESHOLD:([^:]+):', line)
            if num_classifiers == None:
                num_classifiers = int(line)
            elif found_sgram:
                alpha = float(found_sgram.group(1))
                column = found_sgram.group(2)
                value = found_sgram.group(3)
                feature = Feature(alpha, column, value)
                self.features.append(feature)
            elif found_threshold:
                alpha = float(found_threshold.group(1))
                column = found_threshold.group(2)
                feature = Feature(alpha, column)
                self.features.append(feature)
            elif feature != None and len(tokens) == len(self.classes):
                if feature.c0 == []: feature.c0 = [float(x) * feature.alpha for x in tokens]
                elif feature.c1 == []: feature.c1 = [float(x) * feature.alpha for x in tokens]
                elif feature.c2 == []: feature.c2 = [float(x) * feature.alpha for x in tokens]
                else:
                    sys.stderr.write('ERROR: too many weights, in %s, line %d\n' % (shyp_file, line_num))
                    return None
            elif feature != None and feature.value == None and feature.threshold == None and len(tokens) == 1:
                feature.threshold = float(tokens[0])
            else:
                sys.stderr.write('ERROR: unsupported classifier, in %s, line %d\n' % (shyp_file, line_num))
                return None

    # example should be [set(), set(), ...] with a set for each column (or anything which has the "in" operator)
    def compute_scores(self, example):
        #print "rrrr", example
        scores = [0.0 for x in self.classes]
        for feature in self.features:
            if feature.value != None:
                #print feature.value, example[self.names[feature.column]]
                if feature.value in example[self.names[feature.column]]:
                    for i in range(len(feature.c1)):
                        scores[i] += feature.c1[i]
                else:
                    for i in range(len(feature.c0)):
                        scores[i] += feature.c0[i]
            elif feature.threshold != None:
                for value in example[self.names[feature.column]]:
                    if value == '?':
                        #print feature.threshold, value, feature.c0
                        for i in range(len(feature.c0)):
                            scores[i] += feature.c0[i]
                    elif feature.threshold >= float(value):
                        #print feature.threshold, value, feature.c1
                        for i in range(len(feature.c1)):
                            scores[i] += feature.c1[i]
                    else:
                        #print feature.threshold, value, feature.c2
                        for i in range(len(feature.c0)):
                            scores[i] += feature.c2[i]
        return [x / len(self.features) for x in scores]

    def compute_posteriors(self, example):
        scores = self.compute_scores(example)
        #print "eeeee", len(self.features), scores
        return [1.0 / (1.0 + math.exp(-2.0 * x * len(self.features))) for x in scores]

    def classify(self, example):
        scores = self.compute_scores(example)
        max = 0
        argmax = None
        for i in range(len(self.classes)):
            if argmax == None or max < scores[i]:
                max = scores[i]
                argmax = i
        return self.classes[argmax]

def generate_ngram(words, n):
    output = []
    for i in xrange(len(words)):
        current = []
        for j in xrange(n):
            if i + j >= len(words): continue
            current.append(words[i + j])
            output.append('#'.join(current))
    return output

if __name__ == '__main__':
    if len(sys.argv) < 2:
        sys.stderr.write('USAGE: %s <stem>\n' % sys.argv[0])
        sys.exit(1)
    classifier = Classifier(sys.argv[1])
    for line in sys.stdin.xreadlines():
        #print "zzz", line
        columns = line.strip().split(",")
        columns[-1] = re.sub(r'\.$', '', columns[-1].strip())
        scores = classifier.compute_posteriors([set(generate_ngram(x.split(), 3)) for x in columns])
        for i in range(len(classifier.classes)):
            if i != 0:  sys.stdout.write(' ')
            if columns[-1] == classifier.classes[i]:
                sys.stdout.write('1')
            else:
                sys.stdout.write('0')

        #print " ".join([str(x) for x in scores])
        for x in scores:
            sys.stdout.write(" %.12f" % x)
        print ""

